File: src-tauri/src/query/compile_helpers.rs
use std::collections::BTreeMap;

use crate::query::ir::{
    BindValue, FilterGroup, FilterNode, FilterOp, FilterSpec, FilterValue, GroupConnector,
    QuerySpec,
};
use crate::query::validate::ValidatedColumn;
use crate::types::Dialect;

pub(super) fn quote(ident: &str, dialect: Dialect) -> String {
    match dialect {
        Dialect::Postgres => format!("\"{}\"", ident.replace('"', "\"\"")),
        Dialect::MySql => format!("`{}`", ident.replace('`', "``")),
        Dialect::Mssql => format!("[{}]", ident.replace(']', "]]")),
        Dialect::Oracle => format!("\"{}\"", ident.replace('"', "\"\"")),
    }
}

pub(super) fn placeholder(idx: u32, dialect: Dialect) -> String {
    match dialect {
        Dialect::Postgres => format!("${}", idx),
        Dialect::MySql => "?".to_string(),
        Dialect::Mssql => format!("@P{}", idx),
        Dialect::Oracle => format!(":{}", idx),
    }
}

pub(super) fn build_select_clause(
    spec: &QuerySpec,
    dialect: Dialect,
    validated_columns: Option<&[ValidatedColumn]>,
) -> String {
    if let Some(columns) = validated_columns {
        return columns
            .iter()
            .map(|column| {
                format!(
                    "{}.{} AS {}",
                    quote(&column.table_alias, dialect),
                    quote(&column.column, dialect),
                    quote(&column.result_alias, dialect)
                )
            })
            .collect::<Vec<_>>()
            .join(", ");
    }
    if spec.columns.is_empty() {
        return "*".to_string();
    }

    spec.columns
        .iter()
        .map(|c| {
            format!(
                "{}.{}",
                quote(&c.table_alias, dialect),
                quote(&c.column, dialect)
            )
        })
        .collect::<Vec<_>>()
        .join(", ")
}

pub(super) fn build_from_clause(spec: &QuerySpec, dialect: Dialect) -> String {
    if spec.tables.is_empty() {
        return String::new();
    }
    let t = &spec.tables[0];
    format!(
        "{}.{} AS {}",
        quote(&t.schema, dialect),
        quote(&t.name, dialect),
        quote(&t.alias, dialect)
    )
}

pub(super) fn build_join_clause(spec: &QuerySpec, dialect: Dialect) -> String {
    if spec.tables.len() <= 1 {
        return String::new();
    }

    let mut included: std::collections::HashSet<&str> = std::collections::HashSet::new();
    included.insert(spec.tables[0].alias.as_str());

    let mut remaining: Vec<&str> = spec.tables[1..].iter().map(|t| t.alias.as_str()).collect();
    let mut clauses = Vec::new();

    while !remaining.is_empty() {
        let mut found = None;
        for (i, &alias) in remaining.iter().enumerate() {
            for join in &spec.joins {
                let left = join.left_alias.as_str();
                let right = join.right_alias.as_str();
                if left == alias && included.contains(right) {
                    found = Some((i, alias));
                    break;
                } else if right == alias && included.contains(left) {
                    found = Some((i, alias));
                    break;
                }
            }
            if found.is_some() {
                break;
            }
        }

        match found {
            Some((idx, alias)) => {
                remaining.remove(idx);
                included.insert(alias);

                let table_ref = spec.tables.iter().find(|t| t.alias == alias).unwrap();
                let join_target = format!(
                    "INNER JOIN {}.{} AS {}",
                    quote(&table_ref.schema, dialect),
                    quote(&table_ref.name, dialect),
                    quote(&table_ref.alias, dialect)
                );

                let connecting_joins = spec.joins.iter().filter(|candidate| {
                    (candidate.left_alias == alias
                        && included.contains(candidate.right_alias.as_str()))
                        || (candidate.right_alias == alias
                            && included.contains(candidate.left_alias.as_str()))
                });
                let on_clause = connecting_joins
                    .map(|candidate| {
                        format!(
                            "{}.{} = {}.{}",
                            quote(&candidate.left_alias, dialect),
                            quote(&candidate.left_column, dialect),
                            quote(&candidate.right_alias, dialect),
                            quote(&candidate.right_column, dialect)
                        )
                    })
                    .collect::<Vec<_>>()
                    .join(" AND ");

                clauses.push(format!("{} ON {}", join_target, on_clause));
            }
            None => break,
        }
    }

    clauses.join("\n")
}

pub(super) fn build_where_root(
    group: &FilterGroup,
    overrides: &BTreeMap<String, GroupConnector>,
    dialect: Dialect,
    params: &mut Vec<BindValue>,
    param_idx: &mut u32,
) -> Result<String, String> {
    join_children(
        group, overrides, dialect, params, param_idx, /* wrap = */ false,
    )
}

fn join_children(
    group: &FilterGroup,
    overrides: &BTreeMap<String, GroupConnector>,
    dialect: Dialect,
    params: &mut Vec<BindValue>,
    param_idx: &mut u32,
    wrap: bool,
) -> Result<String, String> {
    if group.children.is_empty() {
        return Ok(String::new());
    }

    let mut parts: Vec<(usize, String)> = Vec::with_capacity(group.children.len());
    for (i, child) in group.children.iter().enumerate() {
        let rendered = build_where_node(child, overrides, dialect, params, param_idx)?;
        if !rendered.is_empty() {
            parts.push((i, rendered));
        }
    }

    if parts.is_empty() {
        return Ok(String::new());
    }

    let mut joined = parts[0].1.clone();
    for (orig_i, part) in parts.iter().skip(1) {
        let child = &group.children[*orig_i];
        let connector = overrides
            .get(child_id(child))
            .copied()
            .unwrap_or(group.connector);
        joined.push_str(connector_sql(connector));
        joined.push_str(part);
    }

    if wrap && parts.len() > 1 {
        Ok(format!("({})", joined))
    } else {
        Ok(joined)
    }
}

fn child_id(node: &FilterNode) -> &str {
    match node {
        FilterNode::Leaf(spec) => &spec.id,
        FilterNode::Group(group) => &group.id,
    }
}

fn connector_sql(connector: GroupConnector) -> &'static str {
    match connector {
        GroupConnector::And => " AND ",
        GroupConnector::Or => " OR ",
    }
}

fn build_where_node(
    node: &FilterNode,
    overrides: &BTreeMap<String, GroupConnector>,
    dialect: Dialect,
    params: &mut Vec<BindValue>,
    param_idx: &mut u32,
) -> Result<String, String> {
    match node {
        FilterNode::Leaf(filter) => build_leaf(filter, dialect, params, param_idx),
        FilterNode::Group(group) => build_group(group, overrides, dialect, params, param_idx),
    }
}

fn build_group(
    group: &FilterGroup,
    overrides: &BTreeMap<String, GroupConnector>,
    dialect: Dialect,
    params: &mut Vec<BindValue>,
    param_idx: &mut u32,
) -> Result<String, String> {
    join_children(
        group, overrides, dialect, params, param_idx, /* wrap = */ true,
    )
}

fn build_leaf(
    filter: &FilterSpec,
    dialect: Dialect,
    params: &mut Vec<BindValue>,
    param_idx: &mut u32,
) -> Result<String, String> {
    let column_ref = format!(
        "{}.{}",
        quote(&filter.table_alias, dialect),
        quote(&filter.column, dialect)
    );

    if let Some(op_str) = filter.op.sql_operator() {
        let val = filter.value.as_ref().ok_or_else(|| {
            format!(
                "Filter on {}.{} is missing its value",
                filter.table_alias, filter.column
            )
        })?;
        let lit = match val {
            FilterValue::Single(l) => l,
            _ => return Err(format!("Operator '{}' expects a single value", op_str)),
        };
        let ph = placeholder(*param_idx, dialect);
        *param_idx += 1;
        params.push(BindValue::from_literal(lit)?);
        Ok(format!("{} {} {}", column_ref, op_str, ph))
    } else {
        match filter.op {
            FilterOp::Ilike => {
                let val = filter.value.as_ref().ok_or_else(|| {
                    format!(
                        "Filter on {}.{} is missing its value",
                        filter.table_alias, filter.column
                    )
                })?;
                let lit = match val {
                    FilterValue::Single(l) => l,
                    _ => return Err("ILIKE expects a single value".to_string()),
                };
                let ph = placeholder(*param_idx, dialect);
                *param_idx += 1;
                params.push(BindValue::from_literal(lit)?);
                Ok(build_ilike(&column_ref, &ph, dialect))
            }
            FilterOp::In | FilterOp::NotIn => {
                let val = filter.value.as_ref().ok_or_else(|| {
                    format!(
                        "Filter on {}.{} is missing its value",
                        filter.table_alias, filter.column
                    )
                })?;
                let list = match val {
                    FilterValue::List(l) => l,
                    _ => return Err("IN expects a list of values".to_string()),
                };
                if list.is_empty() {
                    return Ok(if filter.op == FilterOp::In {
                        "1=0".to_string()
                    } else {
                        "1=1".to_string()
                    });
                }
                let mut phs = Vec::with_capacity(list.len());
                for lit in list {
                    let ph = placeholder(*param_idx, dialect);
                    *param_idx += 1;
                    params.push(BindValue::from_literal(lit)?);
                    phs.push(ph);
                }
                let kw = if filter.op == FilterOp::In {
                    "IN"
                } else {
                    "NOT IN"
                };
                Ok(format!("{} {} ({})", column_ref, kw, phs.join(", ")))
            }
            FilterOp::Between => {
                let val = filter.value.as_ref().ok_or_else(|| {
                    format!(
                        "Filter on {}.{} is missing its value",
                        filter.table_alias, filter.column
                    )
                })?;
                let (from, to) = match val {
                    FilterValue::Pair(f, t) => (f, t),
                    _ => return Err("BETWEEN expects a pair of values".to_string()),
                };
                let ph1 = placeholder(*param_idx, dialect);
                *param_idx += 1;
                params.push(BindValue::from_literal(from)?);
                let ph2 = placeholder(*param_idx, dialect);
                *param_idx += 1;
                params.push(BindValue::from_literal(to)?);
                Ok(format!("{} BETWEEN {} AND {}", column_ref, ph1, ph2))
            }
            FilterOp::IsNull => Ok(format!("{} IS NULL", column_ref)),
            FilterOp::IsNotNull => Ok(format!("{} IS NOT NULL", column_ref)),
            FilterOp::IsEmpty => Ok(format!("{} = ''", column_ref)),
            FilterOp::IsNotEmpty => Ok(format!("{} <> ''", column_ref)),
            _ => Err("Unsupported operator in compiler".to_string()),
        }
    }
}

fn build_ilike(column_ref: &str, ph: &str, dialect: Dialect) -> String {
    match dialect {
        Dialect::Postgres => format!("{} ILIKE {}", column_ref, ph),
        Dialect::Mssql => format!("LOWER({}) LIKE LOWER({})", column_ref, ph),
        Dialect::MySql => format!("LOWER({}) LIKE LOWER({})", column_ref, ph),
        Dialect::Oracle => format!("UPPER({}) LIKE UPPER({})", column_ref, ph),
    }
}

File: shared/src/main/kotlin/com/safedb/query/CompileHelpers.kt
package com.safedb.query

import com.safedb.model.BindValue
import com.safedb.model.Dialect
import com.safedb.model.FilterGroup
import com.safedb.model.FilterNode
import com.safedb.model.FilterOp
import com.safedb.model.FilterSpec
import com.safedb.model.FilterValue
import com.safedb.model.GroupConnector
import com.safedb.model.Outcome
import com.safedb.model.QuerySpec
import com.safedb.model.sqlOperator

internal fun quote(ident: String, dialect: Dialect): String = when (dialect) {
    Dialect.Postgres, Dialect.Oracle -> "\"${ident.replace("\"", "\"\"")}\""
    Dialect.MySql -> "`${ident.replace("`", "``")}`"
    Dialect.Mssql -> "[${ident.replace("]", "]]")}]"
}

internal fun placeholder(idx: Int, dialect: Dialect): String = when (dialect) {
    Dialect.Postgres -> "$$idx"
    Dialect.MySql -> "?"
    Dialect.Mssql -> "@P$idx"
    Dialect.Oracle -> ":$idx"
}

internal fun buildSelectClause(
    spec: QuerySpec,
    dialect: Dialect,
    validatedColumns: List<ValidatedColumn>?,
): String {
    if (validatedColumns != null) {
        return validatedColumns.joinToString(", ") { column ->
            "${quote(column.tableAlias, dialect)}.${quote(column.column, dialect)} AS ${quote(column.resultAlias, dialect)}"
        }
    }
    if (spec.columns.isEmpty()) {
        return "*"
    }
    return spec.columns.joinToString(", ") { col ->
        "${quote(col.tableAlias, dialect)}.${quote(col.column, dialect)}"
    }
}

internal fun buildFromClause(spec: QuerySpec, dialect: Dialect): String {
    if (spec.tables.isEmpty()) return ""
    val table = spec.tables[0]
    return "${quote(table.schema, dialect)}.${quote(table.name, dialect)} AS ${quote(table.alias, dialect)}"
}

internal fun buildJoinClause(spec: QuerySpec, dialect: Dialect): String {
    if (spec.tables.size <= 1) return ""

    val included = mutableSetOf(spec.tables[0].alias)
    val remaining = spec.tables.drop(1).map { it.alias }.toMutableList()
    val clauses = mutableListOf<String>()

    while (remaining.isNotEmpty()) {
        var found: Pair<Int, String>? = null
        for ((i, alias) in remaining.withIndex()) {
            for (join in spec.joins) {
                val left = join.leftAlias
                val right = join.rightAlias
                if ((left == alias && included.contains(right)) || (right == alias && included.contains(left))) {
                    found = i to alias
                    break
                }
            }
            if (found != null) break
        }

        val (idx, alias) = found ?: break
        remaining.removeAt(idx)
        included.add(alias)

        val tableRef = spec.tables.first { it.alias == alias }
        val joinTarget = buildString {
            append("INNER JOIN ")
            append(quote(tableRef.schema, dialect))
            append('.')
            append(quote(tableRef.name, dialect))
            append(" AS ")
            append(quote(tableRef.alias, dialect))
        }

        val onClause = spec.joins
            .filter { candidate ->
                (candidate.leftAlias == alias && included.contains(candidate.rightAlias)) ||
                    (candidate.rightAlias == alias && included.contains(candidate.leftAlias))
            }
            .joinToString(" AND ") { candidate ->
                "${quote(candidate.leftAlias, dialect)}.${quote(candidate.leftColumn, dialect)} = " +
                    "${quote(candidate.rightAlias, dialect)}.${quote(candidate.rightColumn, dialect)}"
            }

        clauses.add("$joinTarget ON $onClause")
    }

    return clauses.joinToString("\n")
}

internal fun buildOrderByClause(spec: QuerySpec, dialect: Dialect): String =
    spec.sorts.joinToString(", ") { sort ->
        val direction = when (sort.direction) {
            com.safedb.model.SortDirection.Asc -> "ASC"
            com.safedb.model.SortDirection.Desc -> "DESC"
        }
        "${quote(sort.tableAlias, dialect)}.${quote(sort.column, dialect)} $direction"
    }

internal fun buildGroupByClause(spec: QuerySpec, dialect: Dialect): String =
    spec.groups.joinToString(", ") { group ->
        "${quote(group.tableAlias, dialect)}.${quote(group.column, dialect)}"
    }

internal fun buildWhereRoot(
    group: FilterGroup,
    overrides: Map<String, GroupConnector>,
    dialect: Dialect,
    params: MutableList<BindValue>,
    paramIdx: Int,
): Outcome<Pair<String, Int>> =
    joinChildren(group, overrides, dialect, params, paramIdx, wrap = false)

private fun joinChildren(
    group: FilterGroup,
    overrides: Map<String, GroupConnector>,
    dialect: Dialect,
    params: MutableList<BindValue>,
    paramIdx: Int,
    wrap: Boolean,
): Outcome<Pair<String, Int>> {
    if (group.children.isEmpty()) {
        return Outcome.ok("" to paramIdx)
    }

    var currentIdx = paramIdx
    val parts = mutableListOf<Pair<Int, String>>()
    for ((i, child) in group.children.withIndex()) {
        when (val result = buildWhereNode(child, overrides, dialect, params, currentIdx)) {
            is Outcome.Ok -> {
                currentIdx = result.value.second
                if (result.value.first.isNotEmpty()) {
                    parts.add(i to result.value.first)
                }
            }
            is Outcome.Err -> return Outcome.err(result.message)
        }
    }

    if (parts.isEmpty()) {
        return Outcome.ok("" to currentIdx)
    }

    val joined = buildString {
        append(parts[0].second)
        for ((origI, part) in parts.drop(1)) {
            val child = group.children[origI]
            val connector = overrides[childId(child)] ?: group.connector
            append(connectorSql(connector))
            append(part)
        }
    }

    val rendered = if (wrap && parts.size > 1) "($joined)" else joined
    return Outcome.ok(rendered to currentIdx)
}

private fun childId(node: FilterNode): String = when (node) {
    is FilterNode.Leaf -> node.spec.id
    is FilterNode.Group -> node.group.id
}

private fun connectorSql(connector: GroupConnector): String = when (connector) {
    GroupConnector.And -> " AND "
    GroupConnector.Or -> " OR "
}

private fun buildWhereNode(
    node: FilterNode,
    overrides: Map<String, GroupConnector>,
    dialect: Dialect,
    params: MutableList<BindValue>,
    paramIdx: Int,
): Outcome<Pair<String, Int>> = when (node) {
    is FilterNode.Leaf -> buildLeaf(node.spec, dialect, params, paramIdx)
    is FilterNode.Group -> buildGroup(node.group, overrides, dialect, params, paramIdx)
}

private fun buildGroup(
    group: FilterGroup,
    overrides: Map<String, GroupConnector>,
    dialect: Dialect,
    params: MutableList<BindValue>,
    paramIdx: Int,
): Outcome<Pair<String, Int>> =
    joinChildren(group, overrides, dialect, params, paramIdx, wrap = true)

private fun bindLiteral(literal: com.safedb.model.FilterLiteral): Outcome<BindValue> =
    BindValue.fromLiteral(literal).fold(
        onSuccess = { Outcome.ok(it) },
        onFailure = { Outcome.err(it.message ?: "Invalid literal") },
    )

private fun buildLeaf(
    filter: FilterSpec,
    dialect: Dialect,
    params: MutableList<BindValue>,
    paramIdx: Int,
): Outcome<Pair<String, Int>> {
    val columnRef = "${quote(filter.tableAlias, dialect)}.${quote(filter.column, dialect)}"
    var currentIdx = paramIdx

    if (filter.op.isFriendlyTextPattern()) {
        val literal = (filter.value as? FilterValue.Single)?.literal
            ?: return Outcome.err("${opLabel(filter.op)} expects a single text value")
        val bound = when (val result = bindLiteral(literal)) {
            is Outcome.Ok -> result.value
            is Outcome.Err -> return Outcome.err(result.message)
        }
        val text = (bound as? BindValue.Text)?.value
            ?: return Outcome.err("${opLabel(filter.op)} expects a text value")
        val ph = placeholder(currentIdx, dialect)
        currentIdx += 1
        params.add(BindValue.Text(friendlyPatternText(filter.op, text)))
        val predicate = if (filter.op == FilterOp.ContainsIgnoreCase) {
            buildIlike(columnRef, ph, dialect)
        } else {
            "$columnRef ${friendlyPatternOperator(filter.op)} $ph"
        }
        return Outcome.ok("$predicate ESCAPE '!'" to currentIdx)
    }

    val sqlOp = filter.op.sqlOperator()
    if (sqlOp != null) {
        val value = filter.value
            ?: return Outcome.err("Filter on ${filter.tableAlias}.${filter.column} is missing its value")
        val lit = when (value) {
            is FilterValue.Single -> value.literal
            else -> return Outcome.err("Operator '$sqlOp' expects a single value")
        }
        val ph = placeholder(currentIdx, dialect)
        currentIdx += 1
        when (val bind = bindLiteral(lit)) {
            is Outcome.Ok -> params.add(bind.value)
            is Outcome.Err -> return Outcome.err(bind.message)
        }
        return Outcome.ok("$columnRef $sqlOp $ph" to currentIdx)
    }

    return when (filter.op) {
        FilterOp.Ilike -> {
            val value = filter.value
                ?: return Outcome.err("Filter on ${filter.tableAlias}.${filter.column} is missing its value")
            val lit = when (value) {
                is FilterValue.Single -> value.literal
                else -> return Outcome.err("ILIKE expects a single value")
            }
            val ph = placeholder(currentIdx, dialect)
            currentIdx += 1
            when (val bind = bindLiteral(lit)) {
                is Outcome.Ok -> params.add(bind.value)
                is Outcome.Err -> return Outcome.err(bind.message)
            }
            Outcome.ok(buildIlike(columnRef, ph, dialect) to currentIdx)
        }
        FilterOp.In, FilterOp.NotIn -> {
            val value = filter.value
                ?: return Outcome.err("Filter on ${filter.tableAlias}.${filter.column} is missing its value")
            val list = when (value) {
                is FilterValue.ListValue -> value.literals
                else -> return Outcome.err("IN expects a list of values")
            }
            if (list.isEmpty()) {
                val sql = if (filter.op == FilterOp.In) "1=0" else "1=1"
                return Outcome.ok(sql to currentIdx)
            }
            val phs = mutableListOf<String>()
            for (lit in list) {
                val ph = placeholder(currentIdx, dialect)
                currentIdx += 1
                when (val bind = bindLiteral(lit)) {
                    is Outcome.Ok -> params.add(bind.value)
                    is Outcome.Err -> return Outcome.err(bind.message)
                }
                phs.add(ph)
            }
            val kw = if (filter.op == FilterOp.In) "IN" else "NOT IN"
            Outcome.ok("$columnRef $kw (${phs.joinToString(", ")})" to currentIdx)
        }
        FilterOp.Between -> {
            val value = filter.value
                ?: return Outcome.err("Filter on ${filter.tableAlias}.${filter.column} is missing its value")
            val (from, to) = when (value) {
                is FilterValue.Pair -> value.first to value.second
                else -> return Outcome.err("BETWEEN expects a pair of values")
            }
            val ph1 = placeholder(currentIdx, dialect)
            currentIdx += 1
            when (val bind1 = bindLiteral(from)) {
                is Outcome.Ok -> params.add(bind1.value)
                is Outcome.Err -> return Outcome.err(bind1.message)
            }
            val ph2 = placeholder(currentIdx, dialect)
            currentIdx += 1
            when (val bind2 = bindLiteral(to)) {
                is Outcome.Ok -> params.add(bind2.value)
                is Outcome.Err -> return Outcome.err(bind2.message)
            }
            Outcome.ok("$columnRef BETWEEN $ph1 AND $ph2" to currentIdx)
        }
        FilterOp.IsNull -> Outcome.ok("$columnRef IS NULL" to currentIdx)
        FilterOp.IsNotNull -> Outcome.ok("$columnRef IS NOT NULL" to currentIdx)
        FilterOp.IsEmpty -> Outcome.ok("$columnRef = ''" to currentIdx)
        FilterOp.IsNotEmpty -> Outcome.ok("$columnRef <> ''" to currentIdx)
        else -> Outcome.err("Unsupported operator in compiler")
    }
}

private fun FilterOp.isFriendlyTextPattern(): Boolean = this in setOf(
    FilterOp.Contains,
    FilterOp.ContainsIgnoreCase,
    FilterOp.NotContains,
    FilterOp.StartsWith,
    FilterOp.EndsWith,
)

private fun friendlyPatternOperator(op: FilterOp): String = when (op) {
    FilterOp.NotContains -> "NOT LIKE"
    FilterOp.Contains, FilterOp.StartsWith, FilterOp.EndsWith -> "LIKE"
    else -> error("Not a friendly text pattern: $op")
}

private fun friendlyPatternText(op: FilterOp, text: String): String {
    val escaped = escapeLikeLiteral(text)
    return when (op) {
        FilterOp.Contains, FilterOp.ContainsIgnoreCase, FilterOp.NotContains -> "%$escaped%"
        FilterOp.StartsWith -> "$escaped%"
        FilterOp.EndsWith -> "%$escaped"
        else -> error("Not a friendly text pattern: $op")
    }
}

private fun escapeLikeLiteral(text: String): String = text
    .replace("!", "!!")
    .replace("%", "!%")
    .replace("_", "!_")

private fun buildIlike(columnRef: String, ph: String, dialect: Dialect): String = when (dialect) {
    Dialect.Postgres -> "$columnRef ILIKE $ph"
    Dialect.Mssql, Dialect.MySql -> "LOWER($columnRef) LIKE LOWER($ph)"
    Dialect.Oracle -> "UPPER($columnRef) LIKE UPPER($ph)"
}
